cmd/plc-server/server.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+	"reflect"
+	"strings"
+
+	"github.com/stellentus/go-plc"
+)
+
+type RawTagsHandler struct {
+	plc.ReadWriter
+	validTags map[string]interface{}
+}
+
+func (h RawTagsHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	switch req.Method {
+	case "GET":
+		h.Read(w, req)
+	case "POST":
+		h.Write(w, req)
+	default:
+		logHTTPError(w, http.StatusBadRequest, "Unexpected verb: "+req.Method)
+	}
+}
+
+func logHTTPError(w http.ResponseWriter, code int, reason string) {
+	log.Printf("Bad http request: %s (%d) - %s\n", http.StatusText(code), code, reason)
+	w.WriteHeader(code)
+	w.Write([]byte(reason))
+}
+
+func copy(v interface{}) interface{} {
+	return reflect.New(reflect.TypeOf(v)).Interface()
+}
+
+func (h RawTagsHandler) Read(w http.ResponseWriter, req *http.Request) {
+	tags := req.FormValue("tags")
+	tagsToFetch := map[string]interface{}{}
+
+	if tags != "" {
+		for _, tag := range strings.Split(tags, ",") {
+			v, ok := h.validTags[tag]
+			if !ok {
+				logHTTPError(w, http.StatusNotFound, "Tag does not exist: "+tag)
+				return
+			}
+			tagsToFetch[tag] = copy(v)
```

Comment:
Not a Golang expert here so this might be a silly question: Is the intention here to preallocate the correctly-sized object in advance of line 60?  If so, does the assignment on 60 overwrite the current value of `tagsToFetch[tag]` or does it do something like a "copy constructor" style memcpy?

Reply:
Essentially, yes, but the important line is 59, not 60.
Since the type is `interface{}`, line 59 would end up passing `nil` to `ReadTag`, resulting in an error. We need to create a variable of the correct type to so that `ReadTag` knows what type to read, and also so that there's a variable to read into.
It looks to me like line 58 creates a copy of the `value` inside `tagsToFetch`, but since `tagsToFetch` contains a pointer of the type we want, it's still a reference. So I suspect line 60 might be unnecessary.

Reply:
This is a map, no entry exists until the assignment happens.
The intention is to have the type information in a unique location to avoid race conditions if there are multiple callers.